Modules/posixmodule.c

```diff
@@ -3252,36 +3299,8 @@ posix_getcwd(int use_bytes)
     PyObject *obj;
 
 #ifdef MS_WINDOWS
-    if (!use_bytes) {
-        wchar_t wbuf[MAXPATHLEN];
-        wchar_t *wbuf2 = wbuf;
-        PyObject *resobj;
-        DWORD len;
-        Py_BEGIN_ALLOW_THREADS
-        len = GetCurrentDirectoryW(Py_ARRAY_LENGTH(wbuf), wbuf);
-        /* If the buffer is large enough, len does not include the
-           terminating \0. If the buffer is too small, len includes
-           the space needed for the terminator. */
-        if (len >= Py_ARRAY_LENGTH(wbuf)) {
-            wbuf2 = PyMem_RawMalloc(len * sizeof(wchar_t));
-            if (wbuf2)
-                len = GetCurrentDirectoryW(len, wbuf2);
-        }
-        Py_END_ALLOW_THREADS
-        if (!wbuf2) {
-            PyErr_NoMemory();
-            return NULL;
-        }
-        if (!len) {
-            if (wbuf2 != wbuf)
-                PyMem_RawFree(wbuf2);
-            return PyErr_SetFromWindowsErr(0);
-        }
-        resobj = PyUnicode_FromWideChar(wbuf2, len);
-        if (wbuf2 != wbuf)
-            PyMem_RawFree(wbuf2);
-        return resobj;
-    }
+    if (!use_bytes)
+        return win32_wgetcwd();
```

Comment:
With `win32_wgetcwd()` implemented as a lower-level function, more of the work remains in `posix_getcwd()`, but it's still relatively simple:

```suggestion
    PyObject *obj
    wchar_t *wbuf;
    DWORD len;

    Py_BEGIN_ALLOW_THREADS
    wbuf = win32_wgetcwd(&len);
    Py_END_ALLOW_THREADS

    if (!wbuf) {
        PyErr_NoMemory();
        return NULL;
    }

    obj = PyUnicode_FromWideChar(wbuf, len);
    PyMem_RawFree(wbuf);

    if (use_bytes && obj) {
        Py_SETREF(obj, PyUnicode_EncodeFSDefault(obj));
    }

    return obj;
```